### PR TITLE
feat(tools): Updated get.py with ability to verify extracted files and skip if ok

### PR DESCRIPTION
**Pull request by me-no-dev**
Redo of: https://github.com/espressif/arduino-esp32/pull/8496

New features:

-  Detects corrupted archives (for example not completely downloaded) and automatically re-downloads them (the corrupted archive is renamed with a suffix ".corrupted)
- Detects incomplete extraction (checks only missing filenames - intentional corruption such as changing contents of files is not expected and not detected) - incomplete extraction triggers new and complete extraction
- Updated download report to show download speed and elapsed time
- New parameters:
    - '-v' Verbose - displays more info
    - '-d' Force download all archives (skipping checks)
    - '-e' Force extraction (skipping checks)
    - '-f' Force both download and extraction (skipping checks)
    
Test scenarios (only on Linux)

- Fresh - nothing is downloaded, nothing is extracted
    - Time to complete: 2m51.901s
- Downloaded - only full extraction 
     - Time to complete: 1m22.337s
- Already installed - only performing checks - nothing is missing
    - Time to complete: 0m45.402s
- Multiple corruption - 1 missing archive, 1 archive not downloaded properly, 1 archive not extracted at all, 1 file missing
    - Handled without error or crashing - all tools installed

### CODE REVIEW

**Comment by VojtechBartoska** [comment]
@me-no-dev ready for review?

**Comment by me-no-dev** [comment]
@VojtechBartoska because of how the libs work now and for ease of development, we will probably close this.

**Comment by lucasssvaz** [comment]
Reopened as this will be required to optimize the CI.

**Comment by me-no-dev** [comment]
@lucasssvaz the changes look good to me

**Review by P-R-O-C-H-Y — APPROVED** [review]

File: tools/get.py [hunk 1]
```diff
@@ -254,5 +425,11 @@ def identify_platform():
...
             print("Would install: {0}".format(tool["archiveFileName"]))
         else:
-            get_tool(tool)
+            if not get_tool(tool, force_download, force_extract):
+                if verbose:
+                    print(f"Tool {tool['archiveFileName']} was corrupted. Re-downloading...\n")
+                if not get_tool(
+                    tool, True, force_extract
+                ):  # Corrupted file was renamed, will not be found and will be re-downloaded
+                    print(f"Tool {tool['archiveFileName']} was corrupted, but re-downloading did not help!\n")
```
Comment (lucasssvaz) [inline tools/get.py:434]: Maybe add an `sys.exit(1)` here so the CI will fail when downloading the libs fail

Comment (me-no-dev) [reply tools/get.py:434]: go ahead

Comment (lucasssvaz) [reply tools/get.py:434]: I will. It was mostly a reminder for me to do it when I am back home

**Review by lucasssvaz — APPROVED** [review]

**Comment by me-no-dev** [comment]
@lucasssvaz we need to find out why "Push binary to tools" does not trigger CI

**Comment by lucasssvaz** [comment]
> @lucasssvaz we need to find out why "Push binary to tools" does not trigger CI

It's due to the default behavior of `GITHUB_TOKEN`. In the other CI PR I fixed it (I think at least, it worked in my fork)

**Pull request merged**